src/error.ts

```diff
@@ -0,0 +1,103 @@
+import { inspect } from 'node:util'
+import { Schema, YAMLMap, CollectionTag } from 'yaml'
+import { CreateNodeContext, ToJSContext } from 'yaml/util'
+
+const tag = '!error'
+
+// If the user cared to provide a custom inspect, then use
+// that as the source of extra properties.
+const customInspectSymbol: symbol = inspect.custom
+
+class YAMLErrorObject extends YAMLMap {
+  tag = tag
+  toJSON(_: unknown, ctx: ToJSContext) {
+    ctx = { ...ctx, mapAsMap: false }
+    const { name, message, ...rest } = super.toJSON(_, ctx)
+    const Cls =
+      name === 'EvalError'
+        ? EvalError
+        : name === 'RangeError'
+        ? RangeError
+        : name === 'ReferenceError'
+        ? ReferenceError
+        : name === 'SyntaxError'
+        ? SyntaxError
+        : name === 'TypeError'
+        ? TypeError
+        : name === 'URIError'
+        ? URIError
+        : Error
+    const er = new Cls(message)
+    if (Cls.name !== name) {
+      Object.defineProperty(er, 'name', {
+        value: name,
+        enumerable: false,
+        configurable: true
+      })
+    }
+    if (rest.stack === undefined) er.stack = undefined
+    return Object.assign(er, rest)
+  }
+
+  static from(schema: Schema, value: unknown, ctx: CreateNodeContext) {
+    // Error.cause is standardized, but only recently, so TS does not
+    // yet recognize it as a property on the Error type. It's non-enumerable,
+    // and like stack, very useful to retain in diagnostic information.
+    const error = value as Error & { cause: any }
+    const { name, message, stack, cause } = error
+    const maybeCustom = error as Error & {
+      cause: any,
+      toJSON?: () => any,
+      [customInspectSymbol: symbol]: () => any
+    }
+    const ins =
+      typeof maybeCustom[customInspectSymbol] === 'function' &&
+      maybeCustom[customInspectSymbol]()
+    const json =
+      !(ins && typeof ins === 'object') &&
+      typeof maybeCustom.toJSON === 'function' &&
+      maybeCustom.toJSON()
+    const fields =
+      !!json && typeof json === 'object'
+        ? json
+        : !!ins && typeof ins === 'object'
+        ? ins
+        : { ...error }
+    const withCause = cause === undefined ? {} : { cause }
+
+    // delete the noisy fields that Node.js injects when domains are active
+    delete fields.domain
+    delete fields.domainEmitter
+    delete fields.domainThrew
+    return super.from(
+      schema,
+      {
+        name,
+        message,
+        stack,
+        ...withCause,
```

Comment:
Won't that put `cause: null` in every Error object if it doesn't have a cause property?

Reply:
No, because `undefined` values are left out of the serialization. This is a change in default behaviour between v1 and v2. Admittedly, if someone does have the `keepUndefined` option set, a `cause: null` would indeed be added, so avoiding that with your approach is valid.